### ISSUE TITLE
[GSoC Project Proposal]: Add empirical dynamic model to FIMS

### ISSUE THREAD

**Issue by kellijohnson-NOAA** [ioos/gsoc#107]
### Project Description

The Fisheries Integrated Modeling System (FIMS) is an R package that uses RCPP to allow C++ to work inside of R to assess the status of marine populations that are fished. Currently, we have an age-structured assessment model and we are in the process of adding a surplus production model. It would be amazing to include an empirical dynamic model in the suite of models, which we call model families. The bulk of the code needed can be found in C++ already [rEDM](https://github.com/SugiharaLab/rEDM) but a recent [R package available on GitHub](https://github.com/tanyalrogers/GPEDM) added the Gaussian Process to it. We need a combination of this C++ and R code in our FIMS package.

### Expected Outcomes

The final outcome will be the ability to project a time series forward using EDM within FIMS. Secondarily, there will need to be formal tests added to the package to ensure that the code is implemented appropriately. We use both google tests and testthat to implement tests. The final product will be documentation, including a vignette, on how to use the EDM model within FIMS.

### Skills Required

C++, R

### Expected Project Size

350 hours

### Project Difficulty

Intermediate

### Mentor Name(s)

Andrea Havron, Molly Stevens, Nathan Vaughan, Kelli Johnson

### Mentor GitHub Handle(s)

@andrea-havron-NOAA @mollystevens-noaa @nathanvaughan-NOAA @kellijohnson-NOAA

### Mentor Contact Email(s)

kelli.johnson@noaa.gov, andrea.havron@noaa.gov, nathan.vaughan@noaa.gov, molly.stevens@noaa.gov

### Contributor Evaluation/GSoC GitHub Issue Links

There are several existing [issues in the FIMS repository](https://github.com/NOAA-FIMS/FIMS/issues) that are labeled "good first issue" or "low-hanging fruit" that could be worked on.
- [Good First Issue](https://github.com/NOAA-FIMS/FIMS/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22)
- [Low-hanging 🍎](https://github.com/NOAA-FIMS/FIMS/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22attribute%3A%20low%20hanging%20%F0%9F%8D%8E%22)

### Additional Project Background/Links

Here is a [link to the original issue](https://github.com/NOAA-FIMS/FIMS/issues/1182) in our repository.

**Comment by kellijohnson-NOAA** [comment]
@w-nityammm great to see you are active on IOOS! We have a few other projects planned, one was recently added and a few more will be added tomorrow so keep 👀. Feel free to email me if you have an idea in mind.

**Comment by aman-raj-srivastva** [comment]
Thank you for pointing me to the [surplus-production branch](https://github.com/NOAA-FIMS/FIMS/tree/dev-fix-surplus-production-rebase). I have been reviewing to understand how a new model family is structured within FIMS. It was very helpful to see how the C++ class, interface layer, and Rcpp bindings were added alongside dedicated tests.

Based on this, my current understanding is that an EDM implementation would likely to introduce a new model class in C++ following the existing interface pattern, expose projection functionality through Rcpp bindings and add corresponding unit tests (GoogleTest + testthat)

I also reviewed [Issue #1133](https://github.com/NOAA-FIMS/FIMS/issues/1133) and can see how improving modularity of shared functionality would help ensure the EDM model remains clean and reusable across model families.

For an initial implementation, would it make sense to focus on a minimal working EDM version, and then extend to incorporate Gaussian Process components once the structure is stable?

I’ll continue exploring the branch and documenting a potential integration plan as I go. Please let me know if there are specific design considerations I should prioritize.

**Comment by aman-raj-srivastva** [comment]
Thank you @mollystevens-noaa and @nathanvaughan-NOAA for layered breakdown, this is extremely helpful. Structuring the implementation this way makes a lot of sense from a modularity and long-term extensibility standpoint. It also aligns well with the goal of allowing basic EDM functionality to be reused across model families.

Based on this, I am thinking that a reasonable minimal goal would be:
1. Implementing the base EDM embedding + projection module in C++ within FIMS
2. Exposing a simple forward projection interface via Rcpp
3. Adding full test coverage

Then, in a second phase adding gaussian process weighting as an optional prediction strategy and finally building out the fisheries specific model family following [Tsai et al. (2024)](https://academic.oup.com/icesjms/article/81/7/1209/7696789?login=false) and the [GP-EDM](https://github.com/tanyalrogers/GPEDM).

I will continue reviewing [rEDM](https://github.com/SugiharaLab/rEDM) and [GP-EDM](https://github.com/tanyalrogers/GPEDM) with this layered design in mind and will share a structured draft outline of my proposal for feedback soon.

Also congratulations on the Gulf SSC acceptance 🎉

**Comment by tanyalrogers** [comment]
With the caveat that I haven't used FIMS, and that I'm an ecologist not a computer scientist, here are some additional details that may be helpful.

The structure of GPEDM was not really based on rEDM, which will probably be clear when you dig into the code. The core of GPEDM is based on Steve's MATLAB code, and I designed the front end from scratch. There are some features of the rEDM package that I disliked as a user (e.g. how the training and test data were specified, inability to use the `predict` function and other S3 generics, the default fit statistic being rho, use of special characters in column names) and some that weren't compatible with the GP methodology (e.g. GP has additional hyperparameters that have to be found with an optimizer, the fitting and prediction steps need to be separated because the former is so much more time consuming, the data *have* to be standardized so this is done internally). There are some more notable differences between the packages, like training/test data specification, `tau` being a positive vs. negative number, the how the forecast horizon is handled, and explicit accounting for multiple (replicate) time series. These may be moot points if you're slotting the models into an existing front end, but that is where some of the differences come from.

As far as doing EDM, there are basically 3 steps
1. Constructing delay vectors (aka lag matrix or predictor matrix)
2. Optimize model hyperparameters given predictors, target, and other model specs
3. Given a fitted model, produce predictions for new inputs

Functions in both packages *optionally* combine steps 1 and 2 for convenience only (to quickly run basic models), but they are completely independent steps, and it's important that they be able to be done separately, otherwise it is way too constraining. Forcing them to be separate procedures might not be a bad idea, and increases transparency about what the model predictors and targets are. Making the delay vectors is also not sophisticated, is computationally instantaneous, and probably only needs to be done once, so there isn't any time advantage to it being in C++ vs. R (unless FIMS requires C++). The GPEDM version `makelags` does more stuff than rEDM's `Embed` and can also appropriately handle time series from multiple populations or size classes. The model fitting code should be able to accept any set of predictors and targets; all it's really doing is fitting a nonparametric regression. The use of time lags as predictors is what makes it 'EDM', but really it's completely generic, and it could be used in other ways, e.g. in the same way you'd use a GAM or spline for various applications. The other algorithms (Simplex, S-map) likewise are just different ways of fitting nonparametric regressions that can stand on their own, and could use the same set of predictors and targets.

The model fitting code in GPEDM returns the analytically calculated NLL and gradient, which could be fed to another optimizer than the one used, although it's possible Steve has had better luck with this one that others. I think Tsai's model was built in TMB and produced similar results, so that's an option to explore, although it's not as built out as the version in the R package.

As I mentioned previously, steps 2 and 3 are not really separated in rEDM, in the sense you use the same function for both, and the hyperparameters are fixed, so to find good hyperparameters you have to a grid search and model selection. The parameter theta in S-map could be optimized numerically, and Steve has done it that way, but that's not how it's implemented in rEDM. For the GP, steps 2 and 3 have to be separated because model fitting is slow, while prediction is trivial. This is because a large matrix inversion that has to happen, which isn't present in the other EDM algorithms. I don't know C++, but I included a few simple rcpp functions in GPEDM that I was able to easily look up (and some other dependencies) to try and speed up the most time consuming parts the model, which tend to be matrix algebra. The rest is written in R though because that's the language I know and can thus edit.

When I use rEDM, I tend to use the [legacy version](https://github.com/ha0ye/rEDM), which I've found to be more straightforward and less buggy.

@stevemunch is going to be your expert on the underlying methodology, but if you have specfic questions about the R package implementation, I should be able to answer them.

**Comment by aman-raj-srivastva** [comment]
Thank you @tanyalrogers this explanation really helped clarify the architectural differences and the importance of separating delay construction, model fitting, and prediction. I now better understand how GP-EDM diverges from rEDM and how that impacts integration decisions. I will reflect this separation clearly in my proposal structure and will stay in touch as I refine it.
